cpp/src/community/leiden_kernels.cu

```diff
@@ -0,0 +1,289 @@
+
+    assign_nodes(graph,
+                 delta_Q,
+                 cluster_hash,
+                 src_indices,
+                 next_cluster,
+                 vertex_weights,
+                 cluster_weights,
+                 up_down,
+                 stream);
+
+    up_down = !up_down;
+
+    new_Q = modularity(total_edge_weight, resolution, graph, next_cluster.data().get(), stream);
+
+    if (new_Q > cur_Q) { thrust::copy(next_cluster.begin(), next_cluster.end(), cluster.begin()); }
+  }
+
+  return cur_Q;
+}
+
+template float update_clustering_by_delta_modularity_constrained(
+  float,
+  float,
+  GraphCSRView<int32_t, int32_t, float> const &,
+  rmm::device_vector<int32_t> const &,
+  rmm::device_vector<float> const &,
+  rmm::device_vector<float> &,
+  rmm::device_vector<int32_t> &,
+  rmm::device_vector<int32_t> &,
+  cudaStream_t);
+
+template double update_clustering_by_delta_modularity_constrained(
+  double,
+  double,
+  GraphCSRView<int32_t, int32_t, double> const &,
+  rmm::device_vector<int32_t> const &,
+  rmm::device_vector<double> const &,
+  rmm::device_vector<double> &,
+  rmm::device_vector<int32_t> &,
+  rmm::device_vector<int32_t> &,
+  cudaStream_t);
+
+template <typename vertex_t, typename edge_t, typename weight_t>
+void leiden(GraphCSRView<vertex_t, edge_t, weight_t> const &graph,
+            weight_t &final_modularity,
+            int &num_level,
+            vertex_t *cluster_vec,
+            int max_level,
+            weight_t resolution,
+            cudaStream_t stream)
+{
+#ifdef TIMING
+  HighResTimer hr_timer;
+#endif
+
+  num_level = 0;
+
+  //
+  //  Vectors to create a copy of the graph
+  //
+  rmm::device_vector<edge_t> offsets_v(graph.offsets, graph.offsets + graph.number_of_vertices + 1);
+  rmm::device_vector<vertex_t> indices_v(graph.indices, graph.indices + graph.number_of_edges);
+  rmm::device_vector<weight_t> weights_v(graph.edge_data, graph.edge_data + graph.number_of_edges);
+  rmm::device_vector<vertex_t> src_indices_v(graph.number_of_edges);
+
+  //
+  //  Weights and clustering across iterations of algorithm
+  //
+  rmm::device_vector<weight_t> vertex_weights_v(graph.number_of_vertices);
+  rmm::device_vector<weight_t> cluster_weights_v(graph.number_of_vertices);
+  rmm::device_vector<vertex_t> cluster_v(graph.number_of_vertices);
+
+  //
+  //  Temporaries used within kernels.  Each iteration uses less
+  //  of this memory
+  //
+  rmm::device_vector<vertex_t> tmp_arr_v(graph.number_of_vertices);
+  rmm::device_vector<vertex_t> cluster_inverse_v(graph.number_of_vertices);
+
+  weight_t total_edge_weight =
+    thrust::reduce(rmm::exec_policy(stream)->on(stream), weights_v.begin(), weights_v.end());
+  weight_t best_modularity = -1;
+
+  //
+  //  Initialize every cluster to reference each vertex to itself
+  //
+  thrust::sequence(rmm::exec_policy(stream)->on(stream), cluster_v.begin(), cluster_v.end());
+  thrust::copy(cluster_v.begin(), cluster_v.end(), cluster_vec);
```

Comment:
Consider using `thrust::copy(rmm::exec_policy(stream)->on(stream),...)` that carries stream information.